File: lisp2json.go
package lisp2json

import (
	"encoding/json"
	"fmt"
	"strconv"
	"strings"
)

type LispNode struct {
	Cmd  interface{} `json:"cmd,omitempty"` // Can be string or *LispNode
	Args []LispNode  `json:"args,omitempty"`
	Lit  interface{} `json:"lit,omitempty"`
	Type string      `json:"type,omitempty"`
	Var  string      `json:"var,omitempty"`
}

func Lisp2JSON(input string) (string, error) {
	// Skip lines that start with ;; (Lisp comments)
	var filteredInput strings.Builder
	lines := strings.Split(input, "\n")
	for _, line := range lines {
		trimmedLine := strings.TrimSpace(line)
		if strings.HasPrefix(trimmedLine, ";;") {
			continue // Skip comment lines
		}
		filteredInput.WriteString(line)
		filteredInput.WriteString("\n")
	}

	tokens := tokenize(filteredInput.String())
	var nodes []LispNode
	for len(tokens) > 0 {
		ast, remainingTokens, err := parse(tokens)
		if err != nil {
			return "", err
		}
		nodes = append(nodes, ast)
		tokens = remainingTokens
	}

	jsonBytes, err := json.Marshal(nodes)
	if err != nil {
		return "", err
	}
	return string(jsonBytes), nil
}
func preprocessFunctionSyntax(input string) string {
	var result strings.Builder
	i := 0
	for i < len(input) {
		if strings.HasPrefix(input[i:], "#'(") {
			result.WriteString("(function (")
			i += 3 // Skip "#'("
			parenCount := 1
			closingIndex := i
			for closingIndex < len(input) && parenCount > 0 {
				if input[closingIndex] == '(' {
					parenCount++
				} else if input[closingIndex] == ')' {
					parenCount--
				}
				closingIndex++
			}
			if parenCount == 0 {
				result.WriteString(input[i : closingIndex-1])
				result.WriteString("))")
				i = closingIndex
			} else {
				// If we didn't find a matching closing parenthesis, just write the original syntax
				result.WriteString("#'(")
				i += 3
			}
		} else {
			result.WriteByte(input[i])
			i++
		}
	}
	return result.String()
}
func tokenize(input string) []string {
	var tokens []string
	var currentToken strings.Builder
	inString := false

	// Preprocess #'( syntax
	input = preprocessFunctionSyntax(input)

	// replace all "'(" with (list ;
	input = strings.ReplaceAll(input, "'(", "(list ")

	for _, char := range input {
		if inString {
			currentToken.WriteRune(char)
			if char == '"' {
				tokens = append(tokens, currentToken.String())
				currentToken.Reset()
				inString = false
			}
		} else if char == '"' {
			if currentToken.Len() > 0 {
				tokens = append(tokens, currentToken.String())
				currentToken.Reset()
			}
			currentToken.WriteRune(char)
			inString = true
		} else if char == '(' || char == ')' {
			if currentToken.Len() > 0 {
				tokens = append(tokens, currentToken.String())
				currentToken.Reset()
			}
			tokens = append(tokens, string(char))
		} else if !strings.ContainsRune(" \t\n", char) {
			currentToken.WriteRune(char)
		} else if currentToken.Len() > 0 {
			tokens = append(tokens, currentToken.String())
			currentToken.Reset()
		}
	}

	if currentToken.Len() > 0 {
		tokens = append(tokens, currentToken.String())
	}

	return tokens
}

func parse(tokens []string) (LispNode, []string, error) {
	if len(tokens) == 0 {
		return LispNode{}, tokens, fmt.Errorf("unexpected end of input")
	}

	if tokens[0] == "(" {
		return parseList(tokens[1:])
	} else {
		return parseLiteral(tokens[0]), tokens[1:], nil
	}
}

func parseList(tokens []string) (LispNode, []string, error) {
	if len(tokens) == 0 {
		return LispNode{}, tokens, fmt.Errorf("unexpected end of input")
	}

	if tokens[0] == "let" {
		return parseLet(tokens)
	} else if tokens[0] == "defun" {
		return parseDefun(tokens)
	} else if tokens[0] == "cond" {
		return parseCond(tokens)
	} else if tokens[0] == "lambda" {
		return parseLambda(tokens)
	}

	var args []LispNode
	remaining := tokens

	for len(remaining) > 0 && remaining[0] != ")" {
		arg, newRemaining, err := parse(remaining)
		if err != nil {
			return LispNode{}, tokens, err
		}
		args = append(args, arg)
		remaining = newRemaining
	}

	if len(remaining) == 0 {
		return LispNode{}, tokens, fmt.Errorf("missing closing parenthesis")
	}

	if len(args) == 0 {
		return LispNode{}, remaining[1:], nil
	}

	// Check if the first argument is a function call (like lambda)
	if args[0].Cmd != nil {
		// If the first argument is a function call, create a nested structure
		// where the function is a command that needs to be evaluated first
		return LispNode{
			Cmd:  args[0],
			Args: args[1:],
		}, remaining[1:], nil
	}

	// Otherwise, treat the first argument as a variable name
	return LispNode{Cmd: args[0].Var, Args: args[1:]}, remaining[1:], nil
}
func parseCond(tokens []string) (LispNode, []string, error) {
	if len(tokens) < 2 { // "cond" and at least one clause
		return LispNode{}, tokens, fmt.Errorf("invalid cond expression: not enough arguments")
	}

	// Skip "cond" token
	tokens = tokens[1:]

	var clauses []LispNode
	for len(tokens) > 0 && tokens[0] != ")" {
		if tokens[0] != "(" {
			return LispNode{}, tokens, fmt.Errorf("each cond clause must be a list")
		}
		clause, remaining, err := parseCondClause(tokens[1:])
		if err != nil {
			return LispNode{}, tokens, err
		}
		clauses = append(clauses, clause)
		tokens = remaining
	}

	if len(tokens) == 0 || tokens[0] != ")" {
		return LispNode{}, tokens, fmt.Errorf("missing closing parenthesis for cond expression")
	}

	return LispNode{
		Cmd:  "cond",
		Args: clauses,
	}, tokens[1:], nil
}

func parseCondClause(tokens []string) (LispNode, []string, error) {
	var clause LispNode
	var err error

	// Parse the condition
	clause.Args = append(clause.Args, LispNode{})
	clause.Args[0], tokens, err = parse(tokens)
	if err != nil {
		return LispNode{}, tokens, err
	}

	// Parse the consequent (action)
	for len(tokens) > 0 && tokens[0] != ")" {
		var consequent LispNode
		consequent, tokens, err = parse(tokens)
		if err != nil {
			return LispNode{}, tokens, err
		}
		clause.Args = append(clause.Args, consequent)
	}

	if len(tokens) == 0 || tokens[0] != ")" {
		return LispNode{}, tokens, fmt.Errorf("missing closing parenthesis for cond clause")
	}

	return clause, tokens[1:], nil
}

func parseLet(tokens []string) (LispNode, []string, error) {
	if len(tokens) < 4 { // "let", bindings, body, and closing paren
		return LispNode{}, tokens, fmt.Errorf("invalid let expression: not enough arguments")
	}

	// Skip "let" token
	tokens = tokens[1:]

	// Parse bindings
	bindingsNode, remaining, err := parseBindings(tokens)
	if err != nil {
		return LispNode{}, tokens, fmt.Errorf("error parsing let bindings: %v", err)
	}

	// Parse body (everything else until the closing parenthesis)
	var body []LispNode
	for len(remaining) > 0 && remaining[0] != ")" {
		expr, newRemaining, err := parse(remaining)
		if err != nil {
			return LispNode{}, tokens, fmt.Errorf("error parsing let body: %v", err)
		}
		body = append(body, expr)
		remaining = newRemaining
	}

	// Ensure closing parenthesis
	if len(remaining) == 0 || remaining[0] != ")" {
		return LispNode{}, tokens, fmt.Errorf("missing closing parenthesis for let expression")
	}

	return LispNode{
		Cmd:  "let",
		Args: append([]LispNode{bindingsNode}, body...),
	}, remaining[1:], nil
}

func parseBindings(tokens []string) (LispNode, []string, error) {
	if tokens[0] != "(" {
		return LispNode{}, tokens, fmt.Errorf("bindings must start with '('")
	}
	tokens = tokens[1:] // Skip opening parenthesis

	var bindings []LispNode
	for len(tokens) > 0 && tokens[0] != ")" {
		if tokens[0] != "(" {
			return LispNode{}, tokens, fmt.Errorf("each binding must be a list")
		}
		// Parse a single binding
		tokens = tokens[1:] // Skip opening parenthesis for the binding
		if len(tokens) < 2 {
			return LispNode{}, tokens, fmt.Errorf("each binding must have a variable and a value")
		}

		varName := tokens[0]                           // The variable name (e.g., 'x' or 'y')
		valueNode, remaining, err := parse(tokens[1:]) // The value associated with the variable
		if err != nil {
			return LispNode{}, tokens, err
		}

		// Add the binding as a LispNode where Var is the variable and Args contains the value
		bindings = append(bindings, LispNode{
			Var:  varName,
			Args: []LispNode{valueNode},
		})

		// Expect a closing parenthesis for the current binding
		if remaining[0] != ")" {
			return LispNode{}, tokens, fmt.Errorf("missing closing parenthesis for a binding")
		}
		tokens = remaining[1:] // Skip the closing parenthesis for the current binding
	}

	if len(tokens) == 0 || tokens[0] != ")" {
		return LispNode{}, tokens, fmt.Errorf("missing closing parenthesis for bindings")
	}

	return LispNode{Args: bindings}, tokens[1:], nil
}
func parseLiteral(token string) LispNode {
	if strings.HasPrefix(token, "\"") && strings.HasSuffix(token, "\"") {
		return LispNode{Lit: token[1 : len(token)-1], Type: "string"}
	} else if _, err := strconv.ParseFloat(token, 64); err == nil {
		return LispNode{Lit: token, Type: "number"}
	} else {
		return LispNode{Var: token}
	}
}

func parseDefun(tokens []string) (LispNode, []string, error) {
	if len(tokens) < 4 {
		return LispNode{}, tokens, fmt.Errorf("invalid defun expression: not enough arguments")
	}

	// Skip "defun" token
	tokens = tokens[1:]

	// The first element is the function name
	funcName := tokens[0]
	tokens = tokens[1:]

	// The second element is the argument list (which is a list of variables, not commands)
	if tokens[0] != "(" {
		return LispNode{}, tokens, fmt.Errorf("function argument list must start with '('")
	}
	argListNode, remaining, err := parseArgList(tokens)
	if err != nil {
		return LispNode{}, tokens, err
	}

	// Parse the function body (everything else until the closing parenthesis)
	var body []LispNode
	for len(remaining) > 0 && remaining[0] != ")" {
		expr, newRemaining, err := parse(remaining)
		if err != nil {
			return LispNode{}, tokens, err
		}
		body = append(body, expr)
		remaining = newRemaining
	}

	// Ensure closing parenthesis
	if len(remaining) == 0 || remaining[0] != ")" {
		return LispNode{}, tokens, fmt.Errorf("missing closing parenthesis for defun expression")
	}

	return LispNode{
		Cmd: "defun",
		Args: []LispNode{
			{Var: funcName}, // Function name
			argListNode,     // Argument list
			{Args: body},    // Function body
		},
	}, remaining[1:], nil
}

// Parse argument lists in `defun`, which consist of variables (not commands)
func parseArgList(tokens []string) (LispNode, []string, error) {
	if tokens[0] != "(" {
		return LispNode{}, tokens, fmt.Errorf("argument list must start with '('")
	}
	tokens = tokens[1:] // Skip '('

	var args []LispNode
	for len(tokens) > 0 && tokens[0] != ")" {
		args = append(args, LispNode{Var: tokens[0]}) // Treat each item as a variable
		tokens = tokens[1:]
	}

	if len(tokens) == 0 || tokens[0] != ")" {
		return LispNode{}, tokens, fmt.Errorf("missing closing parenthesis for argument list")
	}

	return LispNode{Args: args}, tokens[1:], nil // Skip closing ')'
}

func (n LispNode) toLisp() string {
	// Handle variables directly
	if n.Var != "" {
		return n.Var
	}

	// Handle literals directly
	if n.Lit != nil {
		if n.Type == "string" {
			return fmt.Sprintf("\"%v\"", n.Lit)
		}
		return fmt.Sprintf("%v", n.Lit)
	}

	// Handle the case where Cmd is a map (nested structure)
	if cmdMap, ok := n.Cmd.(map[string]interface{}); ok {
		// Convert the map to a LispNode
		var innerNode LispNode
		if cmd, ok := cmdMap["cmd"].(string); ok {
			innerNode.Cmd = cmd
		}
		if args, ok := cmdMap["args"].([]interface{}); ok {
			for _, arg := range args {
				if argMap, ok := arg.(map[string]interface{}); ok {
					var argNode LispNode
					if args, ok := argMap["args"].([]interface{}); ok {
						for _, a := range args {
							if aMap, ok := a.(map[string]interface{}); ok {
								var aNode LispNode
								if v, ok := aMap["var"].(string); ok {
									aNode.Var = v
								}
								if l, ok := aMap["lit"].(string); ok {
									aNode.Lit = l
									aNode.Type = aMap["type"].(string)
								}
								if c, ok := aMap["cmd"].(string); ok {
									aNode.Cmd = c
								}
								if aArgs, ok := aMap["args"].([]interface{}); ok {
									for _, aa := range aArgs {
										if aaMap, ok := aa.(map[string]interface{}); ok {
											var aaNode LispNode
											if v, ok := aaMap["var"].(string); ok {
												aaNode.Var = v
											}
											if l, ok := aaMap["lit"].(string); ok {
												aaNode.Lit = l
												aaNode.Type = aaMap["type"].(string)
											}
											aNode.Args = append(aNode.Args, aaNode)
										}
									}
								}
								argNode.Args = append(argNode.Args, aNode)
							}
						}
					}
					innerNode.Args = append(innerNode.Args, argNode)
				}
			}
		}

		innerExpr := innerNode.toLisp()

		// Convert arguments to strings
		args := make([]string, len(n.Args))
		for i, arg := range n.Args {
			args[i] = arg.toLisp()
		}

		return fmt.Sprintf("(%s %s)", innerExpr, strings.Join(args, " "))
	}

	// Handle lambda expressions
	if cmd, ok := n.Cmd.(string); ok && cmd == "lambda" {
		if len(n.Args) < 2 {
			return fmt.Sprintf("(lambda %s ())", n.Args[0].toLisp())
		}
		params := n.Args[0].toLisp()
		body := n.Args[1].toLisp()
		return fmt.Sprintf("(lambda (%s) %s)", params, body)
	}

	// Handle regular function calls
	if cmd, ok := n.Cmd.(string); ok {
		args := make([]string, len(n.Args))
		for i, arg := range n.Args {
			args[i] = arg.toLisp()
		}
		return fmt.Sprintf("(%s %s)", cmd, strings.Join(args, " "))
	}

	// If it's a node without a command (like a list), just join the arguments
	args := make([]string, len(n.Args))
	for i, arg := range n.Args {
		args[i] = arg.toLisp()
	}
	return strings.Join(args, " ")
}

func JSON2Lisp(input string) (string, error) {
	var nodes []LispNode
	err := json.Unmarshal([]byte(input), &nodes)
	if err != nil {
		return "", err
	}

	var lispStrings []string
	for _, node := range nodes {
		lispStrings = append(lispStrings, node.toLisp())
	}

	return strings.Join(lispStrings, "\n"), nil
}

// Helper function to handle let bindings
func (n LispNode) toLispLetBindings() string {
	var bindings []string
	for _, arg := range n.Args {
		if arg.Var != "" && len(arg.Args) > 0 {
			bindings = append(bindings, fmt.Sprintf("(%s %s)", arg.Var, arg.Args[0].toLisp()))
		}
	}
	return fmt.Sprintf("(%s)", strings.Join(bindings, " "))
}

// Parse a lambda expression
func parseLambda(tokens []string) (LispNode, []string, error) {
	if len(tokens) < 4 {
		return LispNode{}, tokens, fmt.Errorf("invalid lambda expression: not enough arguments")
	}

	// Skip "lambda" token
	tokens = tokens[1:]

	// The first element is the argument list (which is a list of variables, not commands)
	if tokens[0] != "(" {
		return LispNode{}, tokens, fmt.Errorf("lambda argument list must start with '('")
	}
	argListNode, remaining, err := parseArgList(tokens)
	if err != nil {
		return LispNode{}, tokens, err
	}

	// Parse the function body (everything else until the closing parenthesis)
	var body []LispNode
	for len(remaining) > 0 && remaining[0] != ")" {
		expr, newRemaining, err := parse(remaining)
		if err != nil {
			return LispNode{}, tokens, err
		}
		body = append(body, expr)
		remaining = newRemaining
	}

	// Ensure closing parenthesis
	if len(remaining) == 0 || remaining[0] != ")" {
		return LispNode{}, tokens, fmt.Errorf("missing closing parenthesis for lambda expression")
	}

	return LispNode{
		Cmd: "lambda",
		Args: []LispNode{
			argListNode,  // Argument list
			{Args: body}, // Function body
		},
	}, remaining[1:], nil
}

File: src/lisp2json.ts
interface LispNode {
  cmd?: string;
  args?: LispNode[];
  lit?: any;
  type?: string;
  var?: string;
}

/**
 * Converts a Lisp expression to JSON
 * @param input The Lisp expression as a string
 * @returns The JSON representation of the Lisp expression
 */
export function lisp2JSON(input: string): string {
  // Skip lines that start with ;; (Lisp comments)
  const filteredInput = input
    .split('\n')
    .filter(line => !line.trim().startsWith(';;'))
    .join('\n');
    
  const tokens = tokenize(filteredInput);
  const nodes: LispNode[] = [];
  
  let remainingTokens = tokens;
  while (remainingTokens.length > 0) {
    const { node, remaining } = parse(remainingTokens);
    nodes.push(node);
    remainingTokens = remaining;
  }

  return JSON.stringify(nodes);
}

/**
 * Converts a JSON representation back to a Lisp expression
 * @param input The JSON representation as a string
 * @returns The Lisp expression
 */
export function json2Lisp(input: string): string {
  const nodes: LispNode[] = JSON.parse(input);
  const lispStrings = nodes.map(node => toLisp(node));
  return lispStrings.join('\n');
}

/**
 * Preprocesses function syntax in the input
 * @param input The input string
 * @returns The preprocessed string
 */
function preprocessFunctionSyntax(input: string): string {
  let result = '';
  let i = 0;
  
  while (i < input.length) {
    if (input.substring(i).startsWith("#'(")) {
      result += "(function (";
      i += 3; // Skip "#'("
      
      let parenCount = 1;
      let closingIndex = i;
      
      while (closingIndex < input.length && parenCount > 0) {
        if (input[closingIndex] === '(') {
          parenCount++;
        } else if (input[closingIndex] === ')') {
          parenCount--;
        }
        closingIndex++;
      }
      
      if (parenCount === 0) {
        result += input.substring(i, closingIndex - 1);
        result += "))";
        i = closingIndex;
      } else {
        // If we didn't find a matching closing parenthesis, just write the original syntax
        result += "#'(";
        i += 3;
      }
    } else {
      result += input[i];
      i++;
    }
  }
  
  return result;
}

/**
 * Tokenizes the input string into an array of tokens
 * @param input The input string
 * @returns An array of tokens
 */
function tokenize(input: string): string[] {
  const tokens: string[] = [];
  let currentToken = '';
  let inString = false;

  // Preprocess #'( syntax
  input = preprocessFunctionSyntax(input);

  // replace all "'(" with (list ;
  input = input.replace(/'\(/g, "(list ");

  for (let i = 0; i < input.length; i++) {
    const char = input[i];
    
    if (inString) {
      currentToken += char;
      if (char === '"') {
        tokens.push(currentToken);
        currentToken = '';
        inString = false;
      }
    } else if (char === '"') {
      if (currentToken.length > 0) {
        tokens.push(currentToken);
        currentToken = '';
      }
      currentToken += char;
      inString = true;
    } else if (char === '(' || char === ')') {
      if (currentToken.length > 0) {
        tokens.push(currentToken);
        currentToken = '';
      }
      tokens.push(char);
    } else if (![' ', '\t', '\n'].includes(char)) {
      currentToken += char;
    } else if (currentToken.length > 0) {
      tokens.push(currentToken);
      currentToken = '';
    }
  }

  if (currentToken.length > 0) {
    tokens.push(currentToken);
  }

  return tokens;
}

/**
 * Parses a list of tokens into a LispNode
 * @param tokens The list of tokens
 * @returns The parsed LispNode and remaining tokens
 */
function parse(tokens: string[]): { node: LispNode; remaining: string[] } {
  if (tokens.length === 0) {
    throw new Error("Unexpected end of input");
  }

  if (tokens[0] === '(') {
    return parseList(tokens.slice(1));
  } else {
    return { node: parseLiteral(tokens[0]), remaining: tokens.slice(1) };
  }
}

/**
 * Parses a list of tokens into a LispNode
 * @param tokens The list of tokens
 * @returns The parsed LispNode and remaining tokens
 */
function parseList(tokens: string[]): { node: LispNode; remaining: string[] } {
  if (tokens.length === 0) {
    throw new Error("Unexpected end of input");
  }

  if (tokens[0] === 'let') {
    return parseLet(tokens);
  } else if (tokens[0] === 'defun') {
    return parseDefun(tokens);
  } else if (tokens[0] === 'cond') {
    return parseCond(tokens);
  }

  const args: LispNode[] = [];
  let remaining = tokens;

  while (remaining.length > 0 && remaining[0] !== ')') {
    const { node, remaining: newRemaining } = parse(remaining);
    args.push(node);
    remaining = newRemaining;
  }

  if (remaining.length === 0) {
    throw new Error("Missing closing parenthesis");
  }

  if (args.length === 0) {
    return { node: {}, remaining: remaining.slice(1) };
  }

  return { node: { cmd: args[0].var, args: args.slice(1) }, remaining: remaining.slice(1) };
}

/**
 * Parses a cond expression
 * @param tokens The list of tokens
 * @returns The parsed LispNode and remaining tokens
 */
function parseCond(tokens: string[]): { node: LispNode; remaining: string[] } {
  if (tokens.length < 2) { // "cond" and at least one clause
    throw new Error("Invalid cond expression: not enough arguments");
  }

  // Skip "cond" token
  tokens = tokens.slice(1);

  const clauses: LispNode[] = [];
  while (tokens.length > 0 && tokens[0] !== ')') {
    if (tokens[0] !== '(') {
      throw new Error("Each cond clause must be a list");
    }
    const { node, remaining } = parseCondClause(tokens.slice(1));
    clauses.push(node);
    tokens = remaining;
  }

  if (tokens.length === 0 || tokens[0] !== ')') {
    throw new Error("Missing closing parenthesis for cond expression");
  }

  return {
    node: {
      cmd: "cond",
      args: clauses
    },
    remaining: tokens.slice(1)
  };
}

/**
 * Parses a cond clause
 * @param tokens The list of tokens
 * @returns The parsed LispNode and remaining tokens
 */
function parseCondClause(tokens: string[]): { node: LispNode; remaining: string[] } {
  const clause: LispNode = { args: [] };
  
  // Parse the condition
  const { node: condition, remaining: afterCondition } = parse(tokens);
  clause.args!.push(condition);
  
  // Parse the consequent (action)
  let remaining = afterCondition;
  while (remaining.length > 0 && remaining[0] !== ')') {
    const { node, remaining: newRemaining } = parse(remaining);
    clause.args!.push(node);
    remaining = newRemaining;
  }

  if (remaining.length === 0 || remaining[0] !== ')') {
    throw new Error("Missing closing parenthesis for cond clause");
  }

  return { node: clause, remaining: remaining.slice(1) };
}

/**
 * Parses a let expression
 * @param tokens The list of tokens
 * @returns The parsed LispNode and remaining tokens
 */
function parseLet(tokens: string[]): { node: LispNode; remaining: string[] } {
  if (tokens.length < 4) { // "let", bindings, body, and closing paren
    throw new Error("Invalid let expression: not enough arguments");
  }

  // Skip "let" token
  tokens = tokens.slice(1);

  // Parse bindings
  const { node: bindingsNode, remaining: afterBindings } = parseBindings(tokens);
  
  // Parse body (everything else until the closing parenthesis)
  const body: LispNode[] = [];
  let remaining = afterBindings;
  
  while (remaining.length > 0 && remaining[0] !== ')') {
    const { node, remaining: newRemaining } = parse(remaining);
    body.push(node);
    remaining = newRemaining;
  }

  // Ensure closing parenthesis
  if (remaining.length === 0 || remaining[0] !== ')') {
    throw new Error("Missing closing parenthesis for let expression");
  }

  return {
    node: {
      cmd: "let",
      args: [bindingsNode, ...body]
    },
    remaining: remaining.slice(1)
  };
}

/**
 * Parses bindings in a let expression
 * @param tokens The list of tokens
 * @returns The parsed LispNode and remaining tokens
 */
function parseBindings(tokens: string[]): { node: LispNode; remaining: string[] } {
  if (tokens[0] !== '(') {
    throw new Error("Bindings must start with '('");
  }
  tokens = tokens.slice(1); // Skip opening parenthesis

  const bindings: LispNode[] = [];
  while (tokens.length > 0 && tokens[0] !== ')') {
    if (tokens[0] !== '(') {
      throw new Error("Each binding must be a list");
    }
    // Parse a single binding
    tokens = tokens.slice(1); // Skip opening parenthesis for the binding
    if (tokens.length < 2) {
      throw new Error("Each binding must have a variable and a value");
    }

    const varName = tokens[0]; // The variable name (e.g., 'x' or 'y')
    const { node: valueNode, remaining } = parse(tokens.slice(1)); // The value associated with the variable

    // Add the binding as a LispNode where Var is the variable and Args contains the value
    bindings.push({
      var: varName,
      args: [valueNode]
    });

    // Expect a closing parenthesis for the current binding
    if (remaining[0] !== ')') {
      throw new Error("Missing closing parenthesis for a binding");
    }
    tokens = remaining.slice(1); // Skip the closing parenthesis for the current binding
  }

  if (tokens.length === 0 || tokens[0] !== ')') {
    throw new Error("Missing closing parenthesis for bindings");
  }

  return { node: { args: bindings }, remaining: tokens.slice(1) };
}

/**
 * Parses a literal token into a LispNode
 * @param token The token to parse
 * @returns The parsed LispNode
 */
function parseLiteral(token: string): LispNode {
  if (token.startsWith('"') && token.endsWith('"')) {
    return { lit: token.slice(1, -1), type: "string" };
  } else if (!isNaN(Number(token))) {
    return { lit: token, type: "number" };
  } else {
    return { var: token };
  }
}

/**
 * Parses a defun expression
 * @param tokens The list of tokens
 * @returns The parsed LispNode and remaining tokens
 */
function parseDefun(tokens: string[]): { node: LispNode; remaining: string[] } {
  if (tokens.length < 4) {
    throw new Error("Invalid defun expression: not enough arguments");
  }

  // Skip "defun" token
  tokens = tokens.slice(1);

  // The first element is the function name
  const funcName = tokens[0];
  tokens = tokens.slice(1);

  // The second element is the argument list (which is a list of variables, not commands)
  if (tokens[0] !== '(') {
    throw new Error("Function argument list must start with '('");
  }
  const { node: argListNode, remaining: afterArgList } = parseArgList(tokens);

  // Parse the function body (everything else until the closing parenthesis)
  const body: LispNode[] = [];
  let remaining = afterArgList;
  
  while (remaining.length > 0 && remaining[0] !== ')') {
    const { node, remaining: newRemaining } = parse(remaining);
    body.push(node);
    remaining = newRemaining;
  }

  // Ensure closing parenthesis
  if (remaining.length === 0 || remaining[0] !== ')') {
    throw new Error("Missing closing parenthesis for defun expression");
  }

  return {
    node: {
      cmd: "defun",
      args: [
        { var: funcName },    // Function name
        argListNode,          // Argument list
        { args: body }        // Function body
      ]
    },
    remaining: remaining.slice(1)
  };
}

/**
 * Parses argument lists in `defun`, which consist of variables (not commands)
 * @param tokens The list of tokens
 * @returns The parsed LispNode and remaining tokens
 */
function parseArgList(tokens: string[]): { node: LispNode; remaining: string[] } {
  if (tokens[0] !== '(') {
    throw new Error("Argument list must start with '('");
  }
  tokens = tokens.slice(1); // Skip '('
  
  const args: LispNode[] = [];
  while (tokens.length > 0 && tokens[0] !== ')') {
    args.push({ var: tokens[0] }); // Treat each item as a variable
    tokens = tokens.slice(1);
  }

  if (tokens.length === 0 || tokens[0] !== ')') {
    throw new Error("Missing closing parenthesis for argument list");
  }
  
  return { node: { args }, remaining: tokens.slice(1) }; // Skip closing ')'
}

/**
 * Converts a LispNode to a Lisp expression string
 * @param node The LispNode to convert
 * @returns The Lisp expression string
 */
function toLisp(node: LispNode): string {
  // Handle variables directly
  if (node.var) {
    return node.var;
  }
  
  // Handle literals directly
  if (node.lit !== undefined) {
    if (node.type === "string") {
      return `"${node.lit}"`; // Return string literals with quotes
    }
    return `${node.lit}`; // Return other literals (like numbers)
  }

  // Handle let expressions
  if (node.cmd === "let") {
    if (!node.args || node.args.length < 2) {
      return "(let ())";  // Handle empty let
    }
    const bindings = toLispLetBindings(node.args[0]);
    const body = toLisp(node.args[1]);
    return `(let ${bindings} ${body})`;
  }

  // Handle defun expressions
  if (node.cmd === "defun") {
    if (!node.args || node.args.length < 3) {
      return `(defun ${toLisp(node.args![0])} ())`;  // Handle empty defun
    }
    const funcName = toLisp(node.args[0]);
    const params = toLisp(node.args[1]);
    const body = toLisp(node.args[2]);
    return `(defun ${funcName} (${params}) ${body})`;
  }

  // Handle function expressions (#'( ... ))
  if (node.cmd === "function") {
    if (node.args && node.args.length === 1) {
      return `#'${toLisp(node.args[0])}`;
    }
    return `#'(${toLisp(node.args![0])})`;
  }

  // Handle list expressions ('( ... ))
  if (node.cmd === "list") {
    if (!node.args) return "()";
    const args = node.args.map(arg => toLisp(arg));
    return `'(${args.join(" ")})`;
  }

  // Handle cond expressions
  if (node.cmd === "cond") {
    if (!node.args) return "(cond)";
    const clauses = node.args.map(clause => {
      if (!clause.args || clause.args.length < 2) {
        return ""; // Skip invalid clauses
      }
      const condition = toLisp(clause.args[0]);
      const consequent = clause.args.slice(1).map(arg => toLisp(arg));
      return `(${condition} ${consequent.join(" ")})`;
    }).filter(Boolean);
    return `(cond ${clauses.join(" ")})`;
  }

  // Handle function calls or expressions
  if (node.cmd) {
    if (!node.args) return `(${node.cmd})`;
    const args = node.args.map(arg => toLisp(arg));
    return `(${node.cmd} ${args.join(" ")})`;
  }
  
  // If it's a node without a command (like a list), just join the arguments
  if (!node.args) return "";
  const args = node.args.map(arg => toLisp(arg));
  return args.join(" ");
}

/**
 * Helper function to handle let bindings
 * @param node The LispNode containing the bindings
 * @returns The Lisp expression string for the bindings
 */
function toLispLetBindings(node: LispNode): string {
  if (!node.args) return "()";
  const bindings = node.args.map(arg => {
    if (arg.var && arg.args && arg.args.length > 0) {
      return `(${arg.var} ${toLisp(arg.args[0])})`;
    }
    return "";
  }).filter(Boolean);
  return `(${bindings.join(" ")})`;
} 